### ISSUE TITLE
Declarative styling

### ISSUE THREAD

**Issue by pjcozzi** [CesiumGS/3d-tiles#2]
# Short-Term
## Spec

**@ggetz**
- [ ] Better name than "3D Tiles Style"
- [ ] Hackathon to experiment with current version

**@pjcozzi**
- [x] TODOs in spec
- [x] Finish spec and schema
## Cesium Implementation

**@ggetz**
- [x] Latest schema for color map and color ramp
  - Remove `BooleanExpression`, `NumberExpression`, and friends
  - Unit tests: `Cesium3DTileStyle` (and implicitly for `Cesium3DTileStyleEngine`)
- [x] Load style from uri, not just JSON string.
- [x] Pass `styleEngine` last to constructors (and make it optional?)
- [x] Example evaluating a style/expression on pick
- [x] Remove chroma - 18d80354be225a116d0a711bb79b190871a82534 and e7aebcddacdcdcc00bb5120215d7f248a5acb29a (and `ColorRampExpression.computeEvenlySpacedIntervals`)

**@pjcozzi**
- [x] Reference doc
- [x] Update CHANGES.md
- [x] Run coverage and see if we are missing any tests for 3D Tiles styles

---
# Later
## Spec
- [ ] Add `JulianDate` datatype to support styling with date/time metadata
- [ ] Introduce translucency property to assign to without having to set RGB?
  - [ ] And/or built-in variables to get the feature's current color and show?
- [x] Operator overloads for regular expressions (we already overload for `Color` so why not here?), see https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/2#issuecomment-184332713
- [ ] Allow variable names inside member expressions, eg. `${foo[${memberName}]}`
- [x] Non-visual properties, e.g., for infobox display, e.g.

``` json
{
    "show" : true,
    "color" : [255, 255, 255],
    "meta" : {
        "description" : "Hello, ${FeatureName}"
    }
}
var str = style.meta.description.evaluate(feature);
```
- [ ] Escape variables in strings.
- [ ] In general, is the ability to assign to feature properties in the style (as part of the JSON or maybe expression) useful?
  - [ ] Improved "standard library"
- [x] Cesium built-in math functions, e.g., `distance`
- [ ] Check out [GLSL](https://www.khronos.org/files/webgl/webgl-reference-card-1_0.pdf)
  - [ ] Built-ins like GLSL, e.g., `czm_time`, `czm_inShadow`, etc.
  - [ ] Include GLSL in a style?
  - [ ] Cascading multiple styles on one tileset?
  - [ ] Add explicit `interval` to optimize `conditional` when used for intervals
  - [ ] Referencing other features in a style

```
color : ramp(palette, distance(${thisFeature.location}, ${anotherFeature.location}))
```

```
show : (${thisFeature.height} > ${anotherBuilding.height})
```
- [ ] Multiple expressions in conditions, https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/2#issuecomment-191754803
- [ ] Generate reference doc from JSON schema
- [ ] Core styling schema + extensions for buildings, vector data, point clouds, etc.?
- [ ] Influence from CZML, i.e., time-dynamic styles
  - [ ] Integration with Cesium `Property` system, e.g., evaluate expressions as `Property` or vice-versa.
- [ ] Other style languages, e.g., more like CSS selectors
- [x] Support vector and matrix properties since this is supported in the batch table binary.
- [ ] Bounding volume hierarchy information (https://github.com/AnalyticalGraphicsInc/cesium/issues/4654)
  - [ ] Tile bounding volumes
  - [ ] Level in tree
- [x] Support vector types like vec2, vec3, and vec4
- [x] Allow component-wise math operations similar to GLSL

## Cesium Implementation
- [ ] Interfaces for `Cesium3DTileStyle` and friends for users writing their own styles.
- [ ] Fix jsep to allow for single backslash characters to be parsed correctly. Workaround is to replace before jsep for now.
- [ ] Column numbers for syntax errors
- [ ] Use regex and replace function for `replaceVariables` in `Expression.js`?
- [x] GPU implementation for tile formats that would really benefit, e.g., point clouds
- [ ] Line width (can we make this as simple as a uniform/batch-value and not recreate geometry?)
- [ ] Expose AST?
- [ ] Expose Conditional properties?
- [ ] Apply to GeoJSON?
- [ ] Optimizations if needed: constant folding and dead code elimination
- [ ] Schema validation - how much do we do at runtime?  Tool for offline validation?
- [ ] Style features in parallel

### Built-in functions

- [x] radians(x) - convert value to radians
- [x] degrees(x) - convert value to degrees
- [x] cos(x)
- [x] sin(x)
- [x] tan(x)
- [x] asin(x)
- [x] acos(x)
- [x] atan(x)
- [x] atan2(y,x) - atan(y,x) in GLSL
- [x] pow(x,y)
- [x] sqrt(x)
- [x] abs(x)
- [x] sign(x)
- [x] floor(x)
- [x] ceil(x)
- [x] round(x)
- [x] min(x,y)
- [x] max(x,y)
- [x] clamp(x,y,z)
- [x] mix(x,y,a)
- [x] length(x)
- [x] distance(x,y)
- [x] dot(x,y)
- [x] cross(x,y)
- [x] normalize(x)
- [ ] random - 0.0 to 1.0 (some complexity on the GLSL side) 
- [x] exp(x)
- [x] log(x)
- [x] exp2(x)
- [x] log2(x)
- [x] fract
- [x] time
- [x] Consider data-driven implementation to avoid lots of duplication where the only difference is the function name and number of arguments.

**Comment by pjcozzi** [comment]
The immediate need for declarative 3D Tiles styling is to be able to write a simple expression that maps a feature's property (or properties) value to its appearance (initially, `show`/`color`), e.g.,
* "Show all buildings greater than 10 meters tall"
* "Use a color ramp to color buildings based on the number of floors"

There's early implementation work in Cesium's [3d-tiles-style](https://github.com/AnalyticalGraphicsInc/cesium/compare/3d-tiles...3d-tiles-style) branch.

The styling format is JSON, and the schema is **not at all final**.  Here's a few examples:
```
// Show all buildings greater than 10 meters tall, e.g., `{Height} > 10`
{
    "show" : {
        "propertyName" : "Height",
        "operator" : ">",
        "operand" : 10
    }
}
```

```
// Create a color ramp  with the provided palette based on the number of floors in a building
{
    "color" : {
        "type" : "ramp",
        "ramp" : {
            "propertyName" : "Floors",
            "colors" : [
                [237, 248, 251],
                [191, 211, 230],
                [158, 188, 218],
                [140, 150, 198],
                [136, 86, 167],
                [129, 15, 124]
            ]
        }
    }
}
```

```
// Map a feature's ID to a color
{
    "color" : {
        "type" : "map",
        "map" : {
            "propertyName" : "id",
            "map" : {
                "1" : [255, 0, 0],
                "2" : [0, 255, 0]
            },
            "default" : [255, 255, 255]
        }
    }
}
```

In Cesium, a style can be applied to a `Cesium3DTileset` object, e.g.,
```javascript
tileset.style = new Cesium.Cesium3DTileStyle(tileset, styleJson);
```
Creating the `Cesium3DTileStyle` object basically "complies" the style, and assigning it to `tileset.style` tells the tileset to apply the style (efficiently based on visibility and if the style or feature properties have changed).

After the style is compiled, it can be changed, e.g.,
```javascript
tileset.style.show.operator = '===';
```
Feature properties that may impact how the style is evaluated can also be changed, e.g.,
```javascript
feature.setProperty('id', 1);
```
In addition to declarative styling, the lower-level API can be used to override the appearance of individual features, e.g.,
```javascript
var feature = scene.pick(movement.endPosition);
if (Cesium.defined(feature)) {
    feature.color = Cesium.Color.YELLOW;
}
```

![img_0350](https://cloud.githubusercontent.com/assets/782098/12847351/be852530-cbe0-11e5-97d4-531a637a9e1f.JPG)

## Schema Ideas

* [ ] Allow properties on both sides of a boolean expression, move from the form `[propertyName] [operator] [operand]` to `[propertyName|operand] [operator] [propertyName|operand]`
   * [ ] Short-term, do we need complex operands, e.g., `Height * 2`
* [ ] Allowed color formats.  Same as CZML or CSS?
* [ ] Explicit intervals for color ramps
* [ ] Support regex for string compares (in boolean expressions and color maps)
* [ ] Maybe: to allow for more concise JSON, move from individual properties to tokenizing a string (probably as simple as a regex), e.g.,
```json
"show" : {
    "propertyName" : "Height",
    "operator" : ">",
    "operand" : 100
}
```
vs.
```json
"show" : "{Height} > 100"
```
* [ ] Or replace explicit expression with more fine-grained primitives like [Tangram filters](https://mapzen.com/documentation/tangram/Filters-Overview/)
* [ ] Compound expressions: &&, ||, !
* [ ] **Invert the design so it isn't color/show focused, but is more like CSS selectors?**
   * Or perhaps this is better served as a separate style format if needed

## Cesium Implementation

* [ ] Additional `ColorExpression` types and interface.
* [ ] Unit tests
* [ ] Reference doc
* [ ] Don't `setEvaluateFunction`/`setIntervals` until all properties change
* [ ] Example evaluating a style/expression on pick
* [ ] Profiling and various `PERFORMANCE_TODO`s
* [ ] How do we evaluate an expression when data comes, for example, from a web service, not the batch table?
* [ ] GPU implementation for tile formats that would really benefit, e.g., point clouds

## 3D Tiles

* [ ] Write spec
* [ ] Write schema

## Think About

* [ ] Translucency
* [ ] Line width (can we make this as simple as a uniform/batch-value and not recreate geometry?)
* [ ] Time-dynamic
* [ ] Integration with CZML, i.e., time-dynamic styles
* [ ] Apply to GeoJSON?
* [ ] Other style languages

## Related-ish

* SQL
* [Tangram Scene file](https://mapzen.com/documentation/tangram/Scene-file/), [filters](https://mapzen.com/documentation/tangram/Filters-Overview/) in particular
* [MapBoxGL Styles](https://www.mapbox.com/mapbox-gl-style-spec/)
* [CartoCSS](https://github.com/mapbox/carto), [selectors](https://www.mapbox.com/tilemill/docs/guides/selectors/) in particular

**Comment by pierotofy** [comment]
I'd much rather use a concise syntax (e.g. "{Height} > 100") than individual properties, which would get really messy for complex rules.

A simple parser can transform the concise syntax into the equivalent structure in a precompilation step.

Something that I could see being used is also the ability to specify styling programmatically, so maybe define an abstract "Cesium3DTileStyleFormatter" interface in Cesium that can be implemented by the developer to specify the styling at runtime, e.g.:

```javascript
var sf = new MyStyleFormatter();
// ...
tileset.style = new Cesium.Cesium3DTileStyle(tileset, sf);
```

Where MyStyleFormatter implements:

```javascript
show: function(tile){
    return tile.Height > 100;
}
```

It could be useful to do thing like:

```javascript
color: function(tile){
   return new StaticColor(Math.Random() * 255, Math.random() * 255, Math.random() * 255);
}
```

To color buildings at randoms (and other possibilities).

**Comment by pjcozzi** [comment]
Thanks for the input, @pierotofy.  We definitely want to support custom functions for expressions, and, I agree, will most likely go with the concise syntax.  Thanks for the code snippets!

**Comment by mramato** [comment]
This is one area where I think Cesium's current Property system can be
leveraged to create some powerful capabilities.  I'm not saying the
Property system exactly as it is today is a perfect fit, but ultimately I
personally feel it's where things will go.  I've wanted to add additional
Property capabilities for a while and this may be a good reason to look
into that.

**Comment by pjcozzi** [comment]
Update https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/2#issuecomment-180355858 to account for the latest implementation work and offline discussion with @mramato.

**Comment by pjcozzi** [comment]
The styling spec will be in the [3d-tiles/spec](https://github.com/AnalyticalGraphicsInc/3d-tiles/compare/style) branch (just a placeholder for now).  I'll bootstrap the prose/schema writing as things start to solidify.

**Comment by pjcozzi** [comment]
**Notes on parsers**

* Expression parsers
   * Small and easy to use: http://jsep.from.so/
   * 3-part tutorial: http://ariya.ofilabs.com/2011/08/math-evaluator-in-javascript-part1.html
      * Library: https://code.google.com/archive/p/tapdigit/
   * More than we need now: http://mathjs.org/
* Full JavaScript parsers
   * Lean: https://github.com/ternjs/acorn
   * Popular: http://jointjs.com/demos/javascript-ast

We'll start with jsep.

**Comment by pjcozzi** [comment]
# Notes on expressions

**Now**

* [ ] Types
   * Number
   * Boolean
   * String
      * RegEx
      * Might be a jsep bug with escaped quotes, e.g., `"a \"b\" c"`
   * Color (CSS color format)
   * Variable (feature property) - resolves to Number, Noolean, String, or Color (or null I suppose)
* [ ] Short-circuit expressions
* [ ] Fast path for well-known expressions, e.g., `${Property} + literal`
* [ ] Examples / test cases (good start, but not complete):
```
"show" : 'a // Syntax error
```
```javascript
"show" : true // Boolean literal
```
```javascript
"show" : !false // Unary NOT
```
```javascript
"show" : true && true // Logical AND
```
```javascript
"show" : false && true // Logical AND, short-circuit
```
```javascript
"show" : false || true // Logical OR
```
```javascript
"show" : true || false // Logical OR, short-circuit
```
```javascript
"show" : (false && false) || (true && true)
```
```javascript
"show" : true ? true : false // Conditional, short-circuit-ish
```
```javascript
"show" : false ? false : true // Conditional, short-circuit-ish
```
```javascript
"show" : 2 > 1 // And similiar for <, <=, >=, ===, !===
```
```javascript
"show" : 0 > -1 // Unary NEGATE
```
```javascript
"show" : (1 + 1) > 1 // *, /, %
```
```javascript
// For now, do not support:
//   * Unary: ~, +
//   * Binary : |, ^, &, ==, !=, <<, >>, >>>
//   * Expressions
//     * Array, e.g., [1, 2]
//     * Compound, e.g., 1; 2; 3
//     * Member, e.g., feature.name // might need this soon
//     * This, e.g., this
// See node types, operations, and literals in the annotated source: http://jsep.from.so/annotated_source/jsep.html
```
```javascript
"show" : "${SomeFeatureProperty}" // Feature property is a boolean
```
```javascript
"show" : "${ZipCode} === 19341" // Variable equals number
```
```javascript
"show" : "${Temperature} > 100"
```
```javascript
"show" : "(${Temperature} > 100) && ((${Weight} / ${Area}) > 2.0)"
```
```javascript
"show" : "${County} === 'Chester'" // Property name equals string
```
```javascript
"show" : "${County} === regExp('/^Chest/')" // String compare with RegEx. Open to other syntax
```
```javascript
"show" : "${County} !== null" // I guess we should support null, what about undefined?
```
```javascript
"show" : 1 // Convert number to boolean I suppose
```
```javascript
// CSS colors
"color" : "#EAA56C"
"color" : "cyan"
"color" : "rgb(100, 255, 190)"
"color" : "hsl(250, 60%, 70%)"
```
```javascript
"color" : "(${Temperature} === 'hot') ? 'red', '#ffffff'"
```
```javascript
"color" : "rgb(255, 0, 0, (${Visible} ? 255 : 0))"
```
```javascript
"color" : "rgb(${red}, ${green}, ${blue}, 255)" // Number properties
```
```javascript
"color" : "rgb(${red} * 255, ${green} * 255, ${blue} * 255, 255)" // Convert 0..1 to 0..255
```

**Later**

* [ ] Replace custom JSON for color ramp/maps with new functions in expressions?
* [ ] Cesium built-in variables, e.g., `czm_*`
* [ ] Typical math functions, e.g., `log`
* [ ] Cesium built-in math functions, e.g., `distance`
* [ ] Optimizations if needed: constant folding and dead code elimination

**Comment by pierotofy** [comment]
For RegEx syntax we could also use (borrowed from Perl/Ruby):

```javascript
"show" : "${County} =~ /^Chest/"
```

null and undefined should probably be both supported (and enforce strong typing):

```javascript
"show" : "${County} !== null && ${Country} !== undefined"
"show" : "${County} !== null" // Does not match undefined values
"show" : "${County} !== undefined" // Does not match null values
```

I would vote in favor of not allowing casts from numbers to bools, just to enforce better code practices, but it's not a big deal if they are allowed.

```javascript
"show" : 1 // Error: "Expected bool, got number"
"show": ${param} === 1 // OK
```

**Comment by pjcozzi** [comment]
Thanks @pierotofy, @ggetz is starting the design/implementation now.  We'll look at these cases; I'm not sure about introducing `=~` yet though as we want to map to JavaScript as best as we can.

**Comment by pjcozzi** [comment]
@ggetz check out the new [Color section](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/style/Styling/README.md#color) in the styling spec and let me know what you think about the TODOs (just open PRs for them) and please make sure our implementation and spec are in-sync and that we reasonably covered the edge cases.

**Comment by ggetz** [comment]
I added my comments in https://github.com/AnalyticalGraphicsInc/3d-tiles/pull/70

**Comment by pjcozzi** [comment]
@ggetz there is a new section on [Number](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/style/Styling/README.md#number).  Let me know your thoughts and please bring our implementation in sync.  Also note that `isNaN()` is `true` and `isFinite()` is `false` due to implicit conversion to Number.

**Comment by pjcozzi** [comment]
@ggetz there is a new section on [conversions](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/style/Styling/README.md#conversions).  Please make sure to test all possible `Color` conversions carefully, e.g., `Color() !== undefined`, `Color() !== null`, `!Color() === false`, etc.  We want the semantics to be the same as if we were using the Cesium `Color` object in JavaScript.

Note that we have to add a `toString()` function to `Color` for implicit conversion to string (or explicit via `Color().toString()`).  We'll have to add a few more functions to make it walk and talk like a real JavaScript object.  More info to follow.

**Comment by pjcozzi** [comment]
@ggetz for strings, make sure we parse both `'` and `"`.

**Comment by pjcozzi** [comment]
@ggetz there is a new section on [operators](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/style/Styling/README.md#operators).  Please make sure our implementation is in-sync (in particular I added the unary `+` and I don't know if we implemented ternary yet) and carefully tested.

**Comment by pjcozzi** [comment]
@ggetz are we in-sync with this:
> Color supports the following binary operators by performing component-wise operations: `===`, `!==`, `+`, `-`, `*`, `/`, and `%`.  For example `Color() === Color()` is true since the red, green, blue, and alpha components are equal.

**Comment by pjcozzi** [comment]
> Note that we have to add a toString() function to Color for implicit conversion to string (or explicit via Color().toString()). We'll have to add a few more functions to make it walk and talk like a real JavaScript object. More info to follow.

I looked more at this.  Just `toString()` is fine for our current purposes.  See Section 19 [here](http://www.ecma-international.org/publications/files/ECMA-ST/Ecma-262.pdf) if interested.

**Comment by pjcozzi** [comment]
@ggetz do you want to design the regex support?

**Comment by ggetz** [comment]
> Color supports the following binary operators by performing component-wise operations: ===, !==, +, -, *, /, and %. For example Color() === Color() is true since the red, green, blue, and alpha components are equal.

Yes this is implemented

**Comment by ggetz** [comment]
> do you want to design the regex support?

Sure

**Comment by ggetz** [comment]
> There is a new section on Number. Let me know your thoughts and please bring our implementation in sync. Also note that isNaN() is true and isFinite() is false due to implicit conversion to Number.

Looks good, I will put this in the implementation

**Comment by ggetz** [comment]
> there is a new section on conversions. Please make sure to test all possible Color conversions carefully, e.g., Color() !== undefined, Color() !== null, !Color() === false, etc. We want the semantics to be the same as if we were using the Cesium Color object in JavaScript.

OK

> there is a new section on operators. Please make sure our implementation is in-sync (in particular I added the unary + and I don't know if we implemented ternary yet) and carefully tested.

Yes, I'll have to add in unary + and ternary.

**Comment by pjcozzi** [comment]
@ggetz there's a new section on [variables](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/style/Styling/README.md#variables).  Please make sure to test all the different datatypes.

**Comment by ggetz** [comment]
> for strings, make sure we parse both ' and ".

I believe jsep handles this, I will add a test to confirm.

**Comment by ggetz** [comment]
> there's a new section on variables. Please make sure to test all the different datatypes.

For variables, do we support multilevel properties? For example, 
```
feature : {
    property : {
        subproperty : 1
    }
}
```
can this be reached by `${property.subproperty}` ?

**Comment by pjcozzi** [comment]
> For variables, do we support multilevel properties? 

Probably, see #65, I just need to double check everything will work out throughout the pipeline.

Hold on this for the moment.

**Comment by pjcozzi** [comment]
@ggetz what is the behavior of CSS colors for values out of range (e.g., `red = 300`).  Please make sure we match it, e.g., throw error, clamp, etc.

**Comment by mramato** [comment]
[Why does HTML think “chucknorris” is a color?](http://stackoverflow.com/questions/8318911/why-does-html-think-chucknorris-is-a-color)

**Comment by ggetz** [comment]
when CSS colors are out of range with the `rgb()`, `hsl()`, etc.. functions, they just clamp the value to the max and min.

**Comment by pjcozzi** [comment]
OK, please make sure we do the same.

Also, what about a string that isn't hex or is too long?

**Comment by ggetz** [comment]
That's covered in the the link @mramato posted.

> make the string a length that is a multiple of 3 by adding 0s: chucknorris0
separate the string into 3 equal length strings: chuc knor ris0
truncate each string to 2 characters: ch kn ri
keep the hex values, and add 0's where necessary: C0 00 00

**Comment by pjcozzi** [comment]
OK, our implementation should work the same.

**Comment by ggetz** [comment]
Trying the behavior myself in chrome, the color is ignored if it is not of the form `#fff` or `#ffffff` or an accepted color keyword, and ignores the color if there are any non-hex charters. Should we go with the legacy behavior described in the link or the chrome behavior?

Also I've been using `Color.fromCssColorString()` to convert the css string. If we go with the former behavior, should I modify the `fromCssColorString` function?

**Comment by mramato** [comment]
We should follow the specification and if `Color.fromCssColorString` deviates from the spec, we should fix it.  MDN has a summary and then links to the actual part of the spec that deals with color: 

MDN: https://developer.mozilla.org/en-US/docs/Web/CSS/color
CSS3 Colors: https://drafts.csswg.org/css-color-3/#color

**Comment by pjcozzi** [comment]
@mramato is right in general, but is @ggetz saying that Chrome implements colors in CSS differently than the spec?

**Comment by ggetz** [comment]
The css3 color spec says:

> The format of an RGB value in hexadecimal notation is a ‘#’ immediately followed by either three or six hexadecimal characters. The three-digit RGB notation (#rgb) is converted into six-digit form (#rrggbb) by replicating digits, not by adding zeros. For example, #fb0 expands to #ffbb00. This ensures that white (#ffffff) can be specified with the short notation (#fff) and removes any dependencies on the color depth of the display.

Which matches the behavior of Chrome (and `Color.fromCssColorString`)

**Comment by pjcozzi** [comment]
So it sounds like everything is in-sync?

**Comment by ggetz** [comment]
Yes.

**Comment by pjcozzi** [comment]
:+1:

**Comment by pjcozzi** [comment]
@ggetz can you bring the implementation in-sync with the latest on dot and bracket notation for member and array access?  See

* [Semantics](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/style/Styling/README.md#semantics)
* [Variables](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/style/Styling/README.md#variables)
* And the new `Color` red, green, blue, and alpha properties

Make sure to test a complex nested use case, e.g., feature properties like:
```json
{
    "aaa" : {
        "bbb" : "ccc",
        "values" : [{
            "ddd" : "eee",
            "ggg" : [1, 2, 3]
        }, {
            "ddd" : "fff",
            "ggg" : [4, 5, 6]
        }]
    }
}
```
(use better names)

**Comment by pjcozzi** [comment]
@ggetz also let me know what you think about this case:

```javascript
var foo = {
  "a.b" : "string",
  "a" : {
    "b" : "object"
  }
};

console.log(foo2.a.b); // object
console.log(foo2['a.b']); // string
```

With style expressions, what does `${a.b}` evaluate to?  Perhaps `object`, and then `${'a.b'}` evaluates to `string`.

We could also require an explicit feature in this case (or always), e.g., `${feature.a.b}` and `${feature['a.b']}`.  Then we need to handle the case when a feature has a feature property, e.g., perhaps `${feature} === ${feature.feature}`

If this is really painful, we could explicitly disallow this for now, give precedence to `object` in this case, and make it a TODO.

**Comment by ggetz** [comment]
`${a.b}` and `${'a.b'}` seem more intuitive and cleaner to me. Requiring an explicit feature without always requiring it is inconsistent. Also by not requiring the explicit feature, we don't need to worry about the `feature` property case.

**Comment by pjcozzi** [comment]
OK with me, see how the implementation feels.

**Comment by pjcozzi** [comment]
@ggetz also please update the spec when you confirm this.

**Comment by ggetz** [comment]
> also please update the spec when you confirm this.

OK

**Comment by pjcozzi** [comment]
@ggetz the spec currently says

> Explicit `Boolean`, `Number`, and `String` constructor functions are not supported.

However, for example, if a tile stores `YearBuilt` as a string, an expression like
```
${YearBuilt} >= 1970
```
"works" because of implicit conversion, but
```
${YearBuilt} === 1970
```
is never true.  Perhaps we should add `Boolean`, `Number`, and `String` for explicit conversion:
```
Number(${YearBuilt}) === 1970
```
If you agree, please update the spec and code.

**Comment by ggetz** [comment]
The unary `+` operator I recently implemented will convert a string to a number. So in this case, you can use `+${YearBuilt} === 1970` to check if this is true. Is that sufficient?

**Comment by pjcozzi** [comment]
For this specific case, yes.  Given that we have omitted `==` and `!=`, I still think we will need these for other cases, but we can wait until we have more experience with styling to see.

**Comment by ggetz** [comment]
OK, I agree these should be added. I will add this to the roadmap.

**Comment by pjcozzi** [comment]
> I will add this to the roadmap.

Also, some useful info is here: http://stackoverflow.com/questions/4719320/new-number-vs-number

**Comment by pjcozzi** [comment]
@ggetz here's some notes for when you are ready to integrate the new expressions into the style JSON schema.

* All the various `show` options go away, and `show` is always a string with an expression that evaluates to Boolean.
* `color` can be a string with an expression that evaluates to Color or an object for a color map or ramp (replacing the [highlighted code here](https://github.com/AnalyticalGraphicsInc/cesium/blob/2e6ab1f675249e87adc25a0570843a08784fd3ae/Apps/Sandcastle/gallery/3D%20Tiles.html#L80-L111)).

Here's a map:
```json
"color" : {
    "key" : "RegEx('^1(\\d)$').exec(${id})",
    "map" : {
        "1" : "Color('#FF0000')",
        "2" : "Color('#00FF00')"
    },
    "defaultValue" : "Color('#FFFFFF')"
}
```
Note that the value in each pair in the map (e.g., `"Color('#FF0000')"`) is an expression evaluating to a Color so it could include variables, etc.

Likewise, I'm pretty sure we will want the key in each pair to also be an expression (kinda like you had before), and we can optimize the case where all keys in the map are literals by building the hash lookup at compile time.

For "color ramps" or "intervals", I think it is best to start with a generic set of conditionals that we execute like a series of `if...else` statements, e.g.,
```json
"color" : {
    "key" : "${Height} / ${Area}",
    "conditional" : {
        "(${KEY} >= 1.0)  && (${KEY} < 10.0)"  : "Color('#FF00FF')",
        "(${KEY} >= 10.0) && (${KEY} < 30.0)"  : "Color('#FF0000')",
        "(${KEY} >= 30.0) && (${KEY} < 50.0)"  : "Color('#FFFF00')",
        "(${KEY} >= 50.0) && (${KEY} < 70.0)"  : "Color('#00FF00')",
        "(${KEY} >= 70.0) && (${KEY} < 100.0)" : "Color('#00FFFF')",
        "(${KEY} >= 100.0)"                    : "Color('#0000FF')",
    }
}
```
`key` could be optional (or even not part of the schema, but it adds a nice readability and easy optimization for complete expressions).  Each conditional property has an expression that evaluates to Boolean and an expression that evaluates to Color.

This approach allows a lot of flexibility and avoids a bunch of schema for closed/open intervals; however, it isn't well suited to a binary search for evaluation (it is basically a linear search).  Optimizing it would require a bunch of analysis; later, if we need, we could just introduce an `interval` schema.

Note that these color maps and ramps could be implemented completely in an expression with the ternary operator, but it would be ugly, painful to generate, and harder to optimize at runtime.

**Comment by ggetz** [comment]
> For "color ramps" or "intervals", I think it is best to start with a generic set of conditionals that we execute like a series of if...else statements

Aren't a map and ramp pretty much the same thing? For example, you can express the "map" in "ramp" syntax pretty easily if we add "defaultValue":

```
"color" : {
    "key" : "RegEx('^1(\\d)$').exec(${id})",
    "conditional" : {
        "${key} === 1" : "Color('#FF0000')",
        "${key} === 2" : "Color('#00FF00')"
    },
    "defaultValue" : "Color('#FFFFFF')"
}
```

**Comment by pjcozzi** [comment]
I actually had that at one point, but made them separate so it would be easy to optimize maps with all literals, e.g., build the hash (JavaScript object) at compile time.  However, I think it is still pretty easy to do the same compile-time analysis with `conditional` (just look for `=== LITERAL` for all properties).  Give it a shot!

**Comment by pjcozzi** [comment]
@ggetz I tightened up the spec in 20d820105127cc3282ca523764c5ae0275547898.  Can you please review?

A few questions:

* Does our `RegExp` type have a `toString()` function like `Color`?  It should.
* We need to explain how operators work with `RegExp` (even if we just disable them for now).  For example, colors say this:

> Colors support the following binary operators by performing component-wise operations: ===, !==, +, -, *, /, and %. For example color() === color() is true since the red, green, blue, and alpha components are equal.

* It's fine to say the following, but we also need to document the arguments like we do for `Color`, `isNaN`, etc.

> The `regExp` function behaves like the JavaScript [`RegExp`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp) constructor and takes the same arguments.

* Can you add a statement to the variables section like 'Variables can be used anywhere an expression can be used except...' or can they be used everywhere (in which case, we should make that explicit)?

**Comment by pjcozzi** [comment]
@ggetz to organize things, I moved our task lists from https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/2#issuecomment-180355858 and https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/2#issuecomment-183918785 to the top comment: https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/2#issue-88423282

**Comment by ggetz** [comment]
> Does our RegExp type have a toString() function like Color? It should. 

Not an explicit function, I will add one.

> We need to explain how operators work with RegExp (even if we just disable them for now). For example, colors say this:

I would say operators do not work with  RegExp at all for now.

I'll add the other items to the spec as well.

**Comment by ggetz** [comment]
Also the changes in `20d8201` look good.

**Comment by pjcozzi** [comment]
@ggetz I stepped through the code in `Expression.js` and made some minor tweaks in https://github.com/AnalyticalGraphicsInc/cesium/commit/10c8c9f914a718cf1660bd7dd1133f6529752966

Here's a few questions/TODOs:

* [ ] Check over all the `try` blocks and only do the part of the code that might throw instead the `try` block.
* [ ] Does `_evaluateLiteralString` need to call `replaceBackslashes`?  Can we do this at compile time instead?
* [ ] Can we do the `instanceof` checks in `_evaluatePlus` at compile time, e.g., so we will end up with more `_evaluate` functions but they will be optimized for their particular type instead of doing type checking at runtime.  If it helps make things clean in the compile step, we could even do a first pass over the AST to compute type info, then a second pass to generate the runtime AST.  Not sure that the extra pass is necessary though.
* [ ] All of the calls to `new Color` (like in `_evaluatePlus`) will be bad for performance.  Can we make `color.evaluate` take a result parameter (see `styleContent` in `Cesium3DTileStyleEngine.js`) and use scratch variables?

**Comment by pjcozzi** [comment]
@ggetz test coverage for `Expression.js` is pretty good at 98%.  Can you please add tests for the `DeveloperError`s that are not covered to bring coverage up to 100%?

**Comment by ggetz** [comment]
> Can we do the instanceof checks in _evaluatePlus at compile time, e.g., so we will end up with more _evaluate functions but they will be optimized for their particular type instead of doing type checking at runtime. 

I'm guessing this would apply to all of the arithmetic `_evaluate` functions?

**Comment by ggetz** [comment]
Also we could do some of the `instanceof` checks at compile time, but we would need to leave one evaluate function with the `instanceof` checks in at runtime because of expressions where the types are not clear until evaluation, for example `${myColor} + 3 * color('red')`

**Comment by pjcozzi** [comment]
> I'm guessing this would apply to all of the arithmetic _evaluate functions?

Yes.

> Also we could do some of the instanceof checks at compile time, but we would need to leave one... 

OK.

**Comment by ggetz** [comment]
Should we wait for the optimized evaluate functions? I believe they are currently on the roadmap under "Later".

**Comment by pjcozzi** [comment]
> Should we wait for the optimized evaluate functions?

Will it be significantly time consuming or significantly deteriorate the code quality?  If not, I would do it now.  If so, then we can evaluate later.

**Comment by ggetz** [comment]
If it's targeted to the operators involving Colors, it shouldn't be too time consuming.

**Comment by pjcozzi** [comment]
@ggetz did you add `key` to https://github.com/AnalyticalGraphicsInc/cesium/pull/3655 as suggested in https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/2#issuecomment-187728554?

For example:

```json
"color" : {
    "key" : "${Height} / ${Area}",
    "conditional" : {
        "(${KEY} >= 1.0)  && (${KEY} < 10.0)"  : "color('#FF00FF')",
        "(${KEY} >= 10.0) && (${KEY} < 30.0)"  : "color('#FF0000')",
        "(${KEY} >= 30.0) && (${KEY} < 50.0)"  : "color('#FFFF00')",
        "(${KEY} >= 50.0) && (${KEY} < 70.0)"  : "color('#00FF00')",
        "(${KEY} >= 70.0) && (${KEY} < 100.0)" : "color('#00FFFF')",
        "(${KEY} >= 100.0)"                    : "color('#0000FF')"
    }
}
```

If not, can you please add it and name it `expression` since `key` came from the map terminology?

**Comment by pjcozzi** [comment]
Also, what do you think of renaming `conditional` in the schema to `conditions`, e.g.,

```json
"color" : {
    "expression" : "RegEx('^1(\\d)$').exec(${id})",
    "conditional" : {
        "{$EXPRESSION} === 1" : "color('#FF0000')",
        "{$EXPRESSION} === 2" : "color('#00FF00')",
        "true" : "color('#FFFFFF')"
    }
}
```
becomes
```json
"color" : {
    "expression" : "RegEx('^1(\\d)$').exec(${id})",
    "conditions" : {
        "{$EXPRESSION} === 1" : "color('#FF0000')",
        "{$EXPRESSION} === 2" : "color('#00FF00')",
        "true" : "color('#FFFFFF')"
    }
}
```

**Comment by pjcozzi** [comment]
> @ggetz did you add key to AnalyticalGraphicsInc/cesium#3655 as suggested in #2 (comment)?

Also, later, we could consider expanding `key`/`expression` into an object that could hold multiple expressions for complex conditions, e.g.,
```
    "color" : {
        "expressions" : {
            "id" : "RegEx('^1(\\d)$').exec(${id})",
            "area" : "${length} * ${height}"
        },
        "conditions" : {
            "({$ID} !== 1) && (${AREA} > 0)" : "color('#FF0000')"
        }
    }
```

This is overkill for now, but let's consider it if experience shows it will be useful.

**Comment by ggetz** [comment]
"Expression" and "conditions"  sound good. For expression, I think `${expression}` is more straight forward than `{$EXPRESSION}` because it looks like the other variables. 

I do like the idea of multiple expressions as well.

**Comment by pjcozzi** [comment]
> I think ${expression} is more straight forward than {$EXPRESSION} because it looks like the other variables.

OK.

Also, this will override a property named `expression` so, in that case, they need to use `feature.expression`.

> I do like the idea of multiple expressions as well.

OK, let's add it when we need it.

**Comment by pjcozzi** [comment]
> I do like the idea of multiple expressions as well.

Also, I already updated the roadmap for this.

**Comment by ggetz** [comment]
> Also, I already updated the roadmap for this.

OK thanks.

**Comment by pjcozzi** [comment]
First cut of the spec is in #80.

**Comment by ggetz** [comment]
> First cut of the spec is in #80.

OK

**Comment by ggetz** [comment]
Suggestions for new name:

* 3D Tiles Ink
* 3D Tiles Flux
* 3D Tiles Chalk
* 3D Tiles Pigment
* 3D Tiles Embellish
* 3D Tiles Flourish

**Comment by twpayne** [comment]
tl;dr The current specification for `conditions` requires JSON object property order to be preserved, which is rarely true.

JavaScript's `for ... in` statement iterates over object properties in an [arbitrary order](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in). The popular Java library [org.json](https://github.com/stleary/JSON-java) defines a JSON object as "[an unordered collection of name/value pairs](https://github.com/stleary/JSON-java/blob/master/JSONObject.java#L46)". Iteration order over [Golang maps](https://blog.golang.org/go-maps-in-action) (a common internal representation for JSON objects) is randomized.

Concrete example: [Styling/README.md](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/master/Styling/README.md) contains the example
```json
{
    "show" : "${Area} > 0",
    "color" : {
        "conditions" : {
            "${Height} < 60" : "color('#13293D')",
            "${Height} < 120" : "color('#1B98E0')",
            "true" : "color('#E8F1F2', 0.5)"
        }
    }
}
```
A feature with property `Height` == 0 matches all three conditions, and therefore its computed color will depend entirely on the evaluation order of the conditions.

Generating `conditions` objects automatically will also be problematic, as some languages and libraries do not allow the order of properties in the serialized JSON output to be controlled.

Suggestion: either require that all `conditions` are mutually exclusive (tricky to enforce) or replace the `conditions` object with an ordered `Array` of `[condition, value]` pairs (uglier, but well defined).

**Comment by pjcozzi** [comment]
Good eye, @twpayne!  I'm surprised we haven't ran into this issue in practice already.

> replace the conditions object with an ordered Array of [condition, value] pairs (uglier, but well defined).

+1 for this.

@lilleyse is it reasonable for you to make this change (code and spec) as part of your current point cloud declarative styling work?

**Comment by lilleyse** [comment]
> @lilleyse is it reasonable for you to make this change (code and spec) as part of your current point cloud declarative styling work?

Yes it should be easy to change. For the spec I'll open a separate PR for this fix. For the code I'll make the change in my styling PR.

**Comment by pjcozzi** [comment]
> Yes it should be easy to change. For the spec I'll open a separate PR for this fix. For the code I'll make the change in my styling PR.

In https://github.com/AnalyticalGraphicsInc/cesium/pull/4336/commits/5cb1088d0dbc8b01d2eb46b289e1ab77150c01b1, part of https://github.com/AnalyticalGraphicsInc/cesium/pull/4336.

**Comment by pjcozzi** [comment]
Labeling `draft 1.0` just for the utility functions and finishing the vector types.

**Comment by pjcozzi** [comment]
@ggetz is there anything here critical for 1.0?

**Comment by ronanmcnulty** [comment]
Is there anyway to color based on whether the lat & lng are within a polygon?

**Comment by ThornM9** [comment]
> Thanks for the input, @pierotofy. We definitely want to support custom functions for expressions, and, I agree, will most likely go with the concise syntax. Thanks for the code snippets!

Were custom functions for expressions ever implemented @pjcozzi ? I'm attempting to color a point cloud by elevation but there's no height attribute for point clouds.

**Comment by pjcozzi** [comment]
@ThornM9 no but there may be another way to achieve your use case, the best place to ask is the Cesium forum: https://groups.google.com/forum/?hl=en#!forum/cesium-dev

**Comment by conradlempert** [comment]
I noticed that atan2(y,x) is more inaccurate than atan(y/x). (While atan(y/x) only works for x > 0 of course). In my particular use case I used the atan2 function to compute the longitude of points in the tileset.

The resulting longitude from atan2(y, x) was off by about 0.00001 (ca. 10 meters), while it looked perfect with atan(y/x). 

Unfortunately I don't know how to tell you the exact resulting numbers since they are only in the GPU I guess, I can only give this estimate from what it looks like. 

Sandcastle:
https://sandcastle.cesium.com/#c=lVX7TyM3EP5XRtFJbErq7CabB081B7SNRAGRcKeTIiFnd0JcHDu1vYSA+N873kcIHHdqfyCsxzPfN283m1daKAcnUmcpTNdwZfSMwRU6NDCQElUDTrTMFlPB4UaJBzRWuDVc66l2IrFwzqcMRglXSqg7DzCQ+DdXqdEwNjpJtBRAR/iLOxRwInTCjRTIJuqBG3gQuCKiI1C4ghO0IluwL7ksmNSS/HyileNCoZnUGvA8UQDkmiEJefogUjT7lWFikEi+aiPTcaES1BsT9VI/mKiCzgmJFt1bvuJf+3RcXAY5R2bkBneo1TVanZkE2czoxcCS2jANom7cijYMRSjMJqiQLY1YCEfZsoynaVDyFo4kWlkHKVonFHdCK3LnnXnCF2g4IxOfjMKjLYP9N+5z4+iLq3bg1QDiOIx6ccy6cRzFnXYUNwp5q9UJOxGLo7DX34uisNMqL+LOXiduhazdi3v9OG63vbieX2ojkNL/HemfyFOq95VwyfxaS1lSd4i022u3+2EnbO/1Or2S4deQtXrdbtiLOu1+e6/f75YXXdbqR+RLv9eJwzDsdyrqsmhl4hhVNl1TwRfCInNzVMEsU4l3a5PbojeK5C7znj6q6s2mOlPe39FyjobSSzGhyQmgbEFGIipYWa/nwjsgIOr2PPgcsvEqp9M+bPRymXhEORJPuA9R2Ni+oSHQr2164k/s+vLb4Pzz+c3ZG81UWD6VeIpLNx9TwU+FdVwlBHlBI0huXl2OhuPhl7Pb4cXvw4vh+Ftl/eI/8qwB+L9mE4Zwr/QK3Jw7+hEW7gUNop6B1OpOuCzFpqTS+g9ycbHMikIDaSrtgEZ9DUuDCeW8UULO9QpJDmudAc08WMQCXzigX4mcsi+1vreVJcw0zd0cX0nfcK2Em5MlV8G6+VhnJc1AWt3wmHauM0l7Ccntmd8uVE/rGcnS064wdyGvahW5H3RD7SlG/2TcYEp9UGb+TEqxtFqk7Osfo37MbrXCSwrnekv7oIJQ2iy4pHKmV2U37WxEwQMm7eDTc1GNy4vbwefR5fnN+OyFPcIvsAO7bzwg4S7sNLx8u9iw8yHC+iOE9f9BePoI4ckj1Os7B9t52hRlQCXwe6iW12JSI+X3CdilW7Zu/vjusT6pHXyI3NpAt36G3fgP2BW+dWuJBFuOYIozeiPs1khSG13jUvIE3/uS99zbyH0DgdNlPyPN4WxGIpq8Cm6jv/8ObjO/1Szt+1Ct+FkWnyiawuylNKdGX3nDT88b9Bc4hhwi2Nr0+s7w5Vwk+UO0Wf1Bvo/qbGNbz7dBnrBqB1YJ+8HLN/LXQa5ULZF8m9QatcNcelzF+ZtYLLVx/okMGGs6XFCWyY/mNEvuiSixtkAAOGxumx6m4gFEevTByw6J5NbSzSyT+Q6d1I4Pm6T/nanU+cvjx1bytVebR8fnhZAxdtik48eWTms55eYd8r8

**Comment by lilleyse** [comment]
See https://github.com/CesiumGS/3d-tiles/issues/179 for optimizing large number of conditional expressions.

**Comment by lilleyse** [comment]
See https://github.com/CesiumGS/3d-tiles/issues/239 for batching vector images into a texture atlas that's part of the payload

**Comment by lilleyse** [comment]
See https://github.com/CesiumGS/3d-tiles/issues/252 for mutables (like GL uniforms)

**Comment by lilleyse** [comment]
See https://github.com/CesiumGS/3d-tiles/issues/266 for styling alpha separately

**Comment by lilleyse** [comment]
See https://github.com/CesiumGS/3d-tiles/issues/268 for accessing feature ids in the styling language

**Comment by lilleyse** [comment]
See https://github.com/CesiumGS/3d-tiles/issues/292 for accessing instanced attributes in the styling language

**Comment by lilleyse** [comment]
More considerations for 3D Tiles Next:

* Can styles access metadata at different granularities including tileset, tile, and group metadata. We support this in CesiumJS.
* Do properties need to be fully qualified based on their class id? e.g. `${classId.propertyId}`? Or only when there are collisions?
* Can [property textures](https://github.com/CesiumGS/glTF/tree/3d-tiles-next/extensions/2.0/Vendor/EXT_mesh_features#property-textures) be styled? Would styling need to happen on the GPU in this case?
* Should geometry attributes like position, normal, etc be available as built-ins and not just for [point clouds](https://github.com/CesiumGS/3d-tiles/tree/main/specification/Styling#point-cloud)?
* Do styles have access to statistics? 
* Do styles have access to class min/max properties values? Both are needed to do color ramps. Statistics are good for localized color ramps, class min/max is good for global color ramps (think a color ramp from the Mariana trench to Everest)
* Do styles have access to enums? `${buildingType} === BuildingType.HOSPITAL`
